tools/common/StaticRegistrar.cs

```diff
@@ -3985,162 +3898,190 @@ void Specialize (AutoIndentStringBuilder sb, ObjCMethod method, List<Exception>
-					setup_return.AppendLine ("xamarin_framework_peer_waypoint ();");
-					setup_return.AppendLine ("mt_dummy_use (retval);");
-				} else {
-					setup_return.AppendLine ("if (!retval) {");
-					setup_return.AppendLine ("res = NULL;");
-					setup_return.AppendLine ("} else {");
+			body.WriteLine ("{");
 
-					if (IsNSObject (type)) {
-						setup_return.AppendLine ("id retobj;");
-						setup_return.AppendLine ("retobj = xamarin_get_nsobject_handle (retval);");
-						setup_return.AppendLine ("xamarin_framework_peer_waypoint ();");
-						setup_return.AppendLine ("[retobj retain];");
-						if (!retain)
-							setup_return.AppendLine ("[retobj autorelease];");
-						setup_return.AppendLine ("mt_dummy_use (retval);");
-						setup_return.AppendLine ("res = retobj;");
-					} else if (IsPlatformType (type, "ObjCRuntime", "Selector")) {
-						setup_return.AppendLine ("res = (SEL) xamarin_get_handle_for_inativeobject (retval, &exception_gchandle);");
-						setup_return.AppendLine ("if (exception_gchandle != INVALID_GCHANDLE) goto exception_handling;");
-					} else if (IsPlatformType (type, "ObjCRuntime", "Class")) {
-						setup_return.AppendLine ("res = (Class) xamarin_get_handle_for_inativeobject (retval, &exception_gchandle);");
-						setup_return.AppendLine ("if (exception_gchandle != INVALID_GCHANDLE) goto exception_handling;");
-					} else if (IsNativeObject (type)) {
-						setup_return.AppendLine ("{0} retobj;", rettype);
-						setup_return.AppendLine ("retobj = xamarin_get_handle_for_inativeobject ((MonoObject *) retval, &exception_gchandle);");
-						setup_return.AppendLine ("if (exception_gchandle != INVALID_GCHANDLE) goto exception_handling;");
-						setup_return.AppendLine ("xamarin_framework_peer_waypoint ();");
-						setup_return.AppendLine ("if (retobj != NULL) {");
-						if (retain) {
-							setup_return.AppendLine ("xamarin_retain_nativeobject (retval, &exception_gchandle);");
-							setup_return.AppendLine ("if (exception_gchandle != INVALID_GCHANDLE) goto exception_handling;");
+			var indent = merge_bodies ? sb.Indentation : sb.Indentation + 1;
+			body.Indentation = indent;
+			body_setup.Indentation = indent;
+			copyback.Indentation = indent;
+			invoke.Indentation = indent;
+			setup_call_stack.Indentation = indent;
+			setup_return.Indentation = indent;
+			cleanup.Indentation = indent;
+
+			// A comment describing the managed signature
+			if (trace) {
+				nslog_start.Indentation = sb.Indentation;
+				comment.Indentation = sb.Indentation;
+				nslog_end.Indentation = sb.Indentation;
+
+				comment.AppendFormat ("// {2} {0}.{1} (", method.Method.DeclaringType.FullName, method.Method.Name, method.Method.ReturnType.FullName);
+				for (int i = 0; i < num_arg; i++) {
+					var param = method.Method.Parameters [i];
+					if (i > 0)
+						comment.Append (", ");
+					comment.AppendFormat ("{0} {1}", param.ParameterType.FullName, param.Name);
```

Comment:
No need to do or own [appendJoin](https://learn.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.appendjoin?view=net-7.0)

Reply:
This was just moving existing code around, so I'll leave this as-is for now.